python/cudf_polars/cudf_polars/dsl/utils/aggregations.py

```diff
@@ -0,0 +1,260 @@
+    name_generator
+        Generator of unique names for temporaries introduced during decomposition.
+    is_top
+        Is this the top of an aggregation expression?
+
+    Returns
+    -------
+    tuple
+        Three-tuple of list of expressions to apply as grouped
+        aggregations (whose children may be evaluated pointwise),
+        single expression to apply to post-process the grouped
+        aggregations, and a boolean indicating whether processing in
+        the inner expression requires aggregations.
+
+    Raises
+    ------
+    NotImplementedError
+        If the expression contains nested aggregations or unsupported
+        operations in a grouped aggregation context.
+    """
+    agg = named_expr.value
+    name = named_expr.name
+    if isinstance(agg, expr.Col):
+        return [named_expr], named_expr, False
+    if isinstance(agg, expr.Len):
+        return [named_expr], expr.NamedExpr(name, expr.Col(agg.dtype, name)), True
+    if isinstance(agg, (expr.Literal, expr.LiteralColumn)):
+        return [], named_expr, False
+    if isinstance(agg, expr.Agg):
+        (child,) = agg.children
+        needs_masking = agg.name in {"min", "max"} and plc.traits.is_floating_point(
+            child.dtype
+        )
+        if needs_masking and agg.options:
+            # pl.col("a").nan_max or nan_min
+            raise NotImplementedError("Nan propagation in groupby for min/max")
+        _, _, has_agg = decompose_single_agg(
+            expr.NamedExpr(next(name_generator), child), name_generator, is_top=False
+        )
+        if has_agg:
+            raise NotImplementedError("Nested aggs in groupby not supported")
+        if needs_masking:
+            child = expr.UnaryFunction(child.dtype, "mask_nans", (), child)
+            # The aggregation is just reconstructed with the new
+            # (potentially masked) child. This is safe because we recursed
+            # to ensure there are no nested aggregations.
+            return (
+                [expr.NamedExpr(name, agg.reconstruct([child]))],
+                expr.NamedExpr(name, expr.Col(agg.dtype, name)),
+                True,
+            )
+        elif agg.name == "sum":
+            col = (
+                expr.Cast(agg.dtype, expr.Col(plc.DataType(plc.TypeId.INT64), name))
+                if (
+                    plc.traits.is_integral(agg.dtype)
+                    and agg.dtype.id() != plc.TypeId.INT64
+                )
+                else expr.Col(agg.dtype, name)
+            )
+            if is_top:
+                # In polars sum(empty_group) => 0, but in libcudf sum(empty_group) => null
+                # So must post-process by replacing nulls, but only if we're a "top-level" agg.
+                rep = expr.Literal(
+                    agg.dtype, pa.scalar(0, type=plc.interop.to_arrow(agg.dtype))
+                )
+                return (
+                    [named_expr],
+                    expr.NamedExpr(
```

Comment:
Side note: https://github.com/rapidsai/cudf/pull/18492 adds `NamedExpr.reconstruct` for this (originally your suggestion).

Reply:
Now used.